service/worker/batcher/workflow.go

```diff
@@ -20,7 +20,328 @@
 
 package batcher
 
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber/cadence/common"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"go.uber.org/cadence"
+	"go.uber.org/cadence/.gen/go/shared"
+	"go.uber.org/cadence/activity"
+	"go.uber.org/cadence/workflow"
+	"golang.org/x/time/rate"
+)
+
 const (
 	batcherContextKey   = "batcherContext"
 	batcherTaskListName = "cadence-sys-batcher-tasklist"
+	batchWFTypeName     = "cadence-sys-batch-workflow"
+	batchActivityName   = "cadence-sys-batch-activity"
+
+	infiniteDuration = 20 * 365 * 24 * time.Hour
+	pageSize         = 1000
+
+	// below are default values for BatchParams
+
+	defaultRPS                      = 50
+	defaultConcurrency              = 5
+	defaultAttemptsOnRetryableError = 50
+	defaultActivityHeartBeatTimeout = time.Minute
 )
+
+const (
+	BatchTypeTerminate = "terminate"
+	BatchTypeReset     = "reset"
+)
+
+type (
+	BatchParams struct {
+		// Target domain to execute batch operation
+		DomainName string
+		// To query the target workflows for processing
+		QueryCondition string
+		// Reaason for the operation
+		Reason string
+		// Supporting: reset,terminate
+		BatchType string
+
+		// Below are all optional
+
+		// RPS of processing. Default to defaultRPS
+		// TODO we will implement smarter way than this static rate limiter: https://github.com/uber/cadence/issues/2138
+		RPS int
+		// Number of goroutines running in parallel to process
+		Concurrency int
+		// Number of attempts for each workflow to process in case of retryable error before giving up
+		AttemptsOnRetryableError int
+		ActivityHeartBeatTimeout time.Duration
+	}
+
+	HeartBeatDetails struct {
+		PageToken   []byte
+		CurrentPage int
+		// This is just an estimation for visibility
+		TotalEstimate int64
```

Comment:
Just curious what is the purpose of this TotalEstimate?

Reply:
When describeWF, you can see the HeartbeatDetails. With this customers can have visibility of the progress.